src/pybind/mgr/dashboard/frontend/src/app/core/auth/user-form/user-form.component.ts

```diff
@@ -0,0 +1,231 @@
+    );
+    this.disableForEdit();
+    this.route.params.subscribe((params: { username: string }) => {
+      const username = params.username;
+      this.userService.get(username).subscribe((userFormModel: UserFormModel) => {
+        this.response = _.cloneDeep(userFormModel);
+        this.setResponse(userFormModel);
+      });
+    });
+  }
+
+  disableForEdit() {
+    this.userForm.get('username').disable();
+  }
+
+  setResponse(response: UserFormModel) {
+    ['username', 'name', 'email', 'roles'].forEach((key) =>
+      this.userForm.get(key).setValue(response[key])
+    );
+  }
+
+  getRequest(): UserFormModel {
+    const userFormModel = new UserFormModel();
+    ['username', 'password', 'name', 'email', 'roles'].forEach(
+      (key) => (userFormModel[key] = this.userForm.get(key).value)
+    );
+    return userFormModel;
+  }
+
+  createAction() {
+    const userFormModel = this.getRequest();
+    this.userService.create(userFormModel).subscribe(
+      () => {
+        this.notificationService.show(
+          NotificationType.success,
+          `User "${userFormModel.username}" has been created.`,
+          'Create User'
+        );
+        this.router.navigate(['/users']);
+      },
+      () => {
+        this.userForm.setErrors({ cdSubmitButton: true });
+      }
+    );
+  }
+
+  editAction() {
+    if (this.isUserRemovingNeededRolePermissions()) {
+      const initialState = {
+        titleText: 'Update user',
+        buttonText: 'Continue',
+        bodyTpl: this.removeSelfUserReadUpdatePermissionTpl,
+        onSubmit: () => {
+          this.modalRef.hide();
+          this.doEditAction();
+        },
+        onCancel: () => {
+          this.userForm.setErrors({ cdSubmitButton: true });
+          this.userForm.get('roles').reset(this.userForm.get('roles').value);
+        }
+      };
+      this.modalRef = this.modalService.show(ConfirmationModalComponent, { initialState });
+    } else {
+      this.doEditAction();
+    }
+  }
+
+  private isCurrentUser(): boolean {
+    return this.authStorageService.getUsername() === this.userForm.get('username').value;
+  }
+
+  private isUserChangingRoles(): boolean {
+    const isCurrentUser = this.isCurrentUser();
+    return (
+      isCurrentUser &&
+      this.response &&
+      !_.isEqual(this.response.roles, this.userForm.getValue('roles'))
+    );
+  }
+
+  private isUserRemovingNeededRolePermissions(): boolean {
+    const isCurrentUser = this.isCurrentUser();
+    return isCurrentUser && !this.hasUserReadUpdatePermissions(this.userForm.get('roles').value);
+  }
+
+  private hasUserReadUpdatePermissions(roles) {
+    for (const role of this.allRoles) {
+      if (roles.indexOf(role.name) !== -1 && role.scopes_permissions['user']) {
+        const userPermissions = role.scopes_permissions['user'];
+        return ['read', 'update'].every((permission) => {
+          return userPermissions.indexOf(permission) !== -1;
```

Comment:
`return userPermission[permission]` would be as sufficient, as it would return `undefined` if not there which would resolve to `false`.

Reply:
Note that `userPermission` is an array, so your suggestion does not work here, because

````
['read', 'update'].every((permission) => { 
  return ["create", "delete", "read", "update"].indexOf(permission) !== -1 
});
````
returns `true`, while:

````
['read', 'update'].every((permission) => { 
  return ["create", "delete", "read", "update"][permission]
});
````
returns `false`.